query/physicalplan/binaryscalarexpr.go

```diff
@@ -38,6 +50,221 @@ type BinaryScalarExpr struct {
+			}
+			return nil
+		})
+		return bm, col, err
+	case logicalplan.OpGt:
+		col, err := forEachParquetValue(left, func(i int, value parquet.Value) error {
+			if ParquetValueCompareArrowScalar(value, right) > 0 {
+				bm.AddInt(i)
+			}
+			return nil
+		})
+		return bm, col, err
+	case logicalplan.OpGtEq:
+		col, err := forEachParquetValue(left, func(i int, value parquet.Value) error {
+			if ParquetValueCompareArrowScalar(value, right) >= 0 {
+				bm.AddInt(i)
+			}
+			return nil
+		})
+		return bm, col, err
+	case logicalplan.OpRegexMatch, logicalplan.OpRegexNotMatch:
+		col, err := forEachParquetValue(left, func(i int, value parquet.Value) error {
+			match, err := regexp.MatchString(right.String(), value.String())
+			if err != nil {
+				return err
+			}
+			if match && operator == logicalplan.OpRegexMatch || !match && operator == logicalplan.OpRegexNotMatch {
+				bm.AddInt(i)
+			}
+			return nil
+		})
+		return bm, col, err
+	default:
+		return nil, nil, fmt.Errorf("unsupported operator: %v", operator)
+	}
+}
+
+// ParquetValueCompareArrowScalar compares a parquet.Value to a scalar.Scalar
+// It returns 0 if they are equal, -1 if the parquet.Value is less than the scalar.Scalar, and 1 if the parquet.Value is greater than the scalar.Scalar.
+func ParquetValueCompareArrowScalar(v parquet.Value, s scalar.Scalar) int {
+	switch v.Kind() {
+	case parquet.Boolean:
+		if v.Boolean() == s.(*scalar.Boolean).Value {
+			return 0
+		} else if v.Boolean() {
+			return 1
+		} else {
+			return -1
+		}
+	case parquet.Int32:
+		if v.Int32() == s.(*scalar.Int32).Value {
+			return 0
+		} else if v.Int32() > s.(*scalar.Int32).Value {
+			return 1
+		} else {
+			return -1
+		}
+	case parquet.Int64:
+		if v.Int64() == s.(*scalar.Int64).Value {
+			return 0
+		} else if v.Int64() > s.(*scalar.Int64).Value {
+			return 1
+		} else {
+			return -1
+		}
+	case parquet.Float:
+		if v.Float() == s.(*scalar.Float32).Value {
+			return 0
+		} else if v.Float() > s.(*scalar.Float32).Value {
+			return 1
+		} else {
+			return -1
+		}
+	case parquet.Double:
+		if v.Double() == s.(*scalar.Float64).Value {
+			return 0
+		} else if v.Double() > s.(*scalar.Float64).Value {
+			return 1
+		} else {
+			return -1
+		}
+	case parquet.ByteArray:
+		if string(v.ByteArray()) == s.(*scalar.String).String() {
+			return 0
+		} else if string(v.ByteArray()) > s.(*scalar.String).String() {
+			return 1
+		} else {
+			return -1
+		}
+	default:
+		panic(fmt.Sprintf("unsupported type: %v", v.Kind()))
+	}
+}
+
+func forEachParquetValue(chunk parquet.ColumnChunk, f func(i int, value parquet.Value) error) ([]parquet.Value, error) {
+	vals := make([]parquet.Value, chunk.NumValues())
+	pages := chunk.Pages()
+	defer pages.Close()
+	i := 0
+	n := 0
+	for {
+		p, err := pages.ReadPage()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+		reader := p.Values()
+		n, err = reader.ReadValues(vals[n:])
+		if err != nil && err != io.EOF {
```

Comment:
What happens if an `io.EOF` is returned? Do we continue the loop and `break` in the next iteration?

Reply:
That's correct. We'll process whatever values may/may not have been returned and will hit EOF on the next iteration.